File: 第六章_二分搜索/二分搜索(2)/快速N次方练习题/QuickPower.cpp
const int MOD=1000000007;
  
class QuickPower {
public:
    int getPower(int k, int N) {
              
        long res=1;
        long temp=k;
        for(;N>0;N>>=1)// 右移一位
        {
            if(N&1==1)// 位运算 若不等于0 说明此时i的二进制末尾为1 要进行计算
            {
                res*=temp;
            }
            temp=(temp*temp)%MOD;
            res%=MOD;
        }
  
        return res;
    }
};

File: 第六章_二分搜索/二分搜索(2)/快速N次方练习题/QuickPower.java
public class QuickPower {
	public int getPower(int k, int N) {

		if (N == 0) {
			return 1;
		}
		long res = 1;
		long temp = k;
		for (int i = N; i > 0; i >>= 1) {// 右移一位
			if ((i & 1) != 0) {// 位运算 若不等于0 说明此时i的二进制末尾为1 要进行计算
				res *= temp;
			}
			temp = (temp * temp) % 1000000007;
			res = res % 1000000007;
		}

		return (int) res;
	}
}